### DISCUSSION TITLE
Mensaje de advertencia notificado como error

### DISCUSSION THREAD

**Discussion by daniel89fg** [Barnetik/tbai-php-lib#26] [Q&A] [answered]
Hola.

Soy el programador del plugin Ticketbai para FacturaScripts. Usamos la librería para generar el xml mediante json. Hasta aquí todo bien. Al enviar los datos a la diputación de Áraba (con Guipúzcoa no pasa) la factura se firma bien, se recibe un código 200 y un mensaje de advertencia.

Al hacer la petición a hacienda la respuesta recibida compruebo que si el código es del estado es 200, y si las funciones isDelivered() e isCorrect() son falsas, muestro el mensaje y termino. Pero en realidad debería de mostrar el mensaje y continuar por que si que se firmo y el estado fue 200.

La pregunta quizás es por que en Áraba al recibir un mensaje de advertencia con el código 009, las funciones isDelivered() e isCorrect() son falsas? en cambio la misma factura enviada a Guipúzcoa no devuelve dicho mensaje.

- ¿Debería la librería comprobar dicho código de advertencia y actuar en consecuencia?
- ¿Las funciones isDelivered() e isCorrect() que comprueban internamente?

**Comment by arraintxo** [comment]
Buenas @daniel89fg ,

la función `isDelivered` devuelve false en caso de que el código de Estado de la respuesta sea diferente a `'00'`, ya que si la factura se ha registrado correctamente en el sistema TicketBai, ese es el código devuelto en el XML:
![irudia](https://user-images.githubusercontent.com/1630107/232801971-de7f2e3d-1b07-4ef7-bfb8-00ff460db45d.png)

A pesar de devolver un 200 como estado de HTTP, la factura no se ha registrado correctamente.

Por su parte la función `isCorrect` devuelve `false` siempre que el isDelivered devuelve `false`, o en los casos en los que a pesar de haberse registrado la factura (`isDelivered: true`) esta ha devuelto mensajes de error en las validaciones.

Para saber cual es el error, deberías poder recuperar la lista de errores utilizando el método `mainErrorMessage` que devuelve un objeto JSON con los errores detectados (tal vez sería mejor tener otro método que devuelva directamente el array de errores de validación).

Respecto al código 009, se trata de un error de encadenamiento:
![irudia](https://user-images.githubusercontent.com/1630107/232804284-4ade614d-97ad-41f1-8e9a-938a2d34c78d.png)

**Comment by daniel89fg** [comment]
El problema es que creamos una nueva serie, creamos la 1º factura de dicha serie. Enviamos a firmar la factura a la diputación de Araba y recibimos el error 009, "error de encadenamiento", no entendemos como es posible si es la 1º factura, no existen facturas anteriores. En cambio hacemos el mismo proceso pero enviando a la diputación de Guipúzcoa y todo correcto.

Además, aunque dio error 009 en Araba, si que se registro.

Quizás lo que debería de hacer es lo siguiente.

- Comprobar si $response->status() es false, en tal caso terminar y mostrar el mensaje de error recibido. Si $response->status() es true, continuamos.
- Comprobar si $response->isDelivered() es false o $response->isCorrect() es false, en tal caso mostrar el mensaje de error recibido pero permitir continuar.

¿correcto?, esto sería lo ideal.

**Comment by arraintxo** [comment]
La verdad es que es raro que Araba y Gipuzkoa funcionen de manera diferente en este aspecto.

Que devuelva un 009 no debería ser un problema cuándo se trata de la primera factura del emisor, pero no tengo claro como se debe proceder al cambiar de serie, cuando el emisor ya existe y el origen es del mismo software.

Estoy mirando lo que ocurre con el isDelivered y el código devuelto por las APIs de las diputaciones, había dado por hecho que un "Rechazado" (01) significaba que la factura no se registra (que parecía lo más lógico), pero parece que al menos en el caso de Araba esta sí se registra, aunque tampoco sé si se registra en todos los casos (no sé si depende de la lista de errores devueltos). Habría que investigar un poco más.

Mientras, parece que podrías mirar si el isDelivered es false y en caso de serlo comprobar si el status es distinto de 200... En ese caso seguro que ha habido un error de comunicación.

**Comment by daniel89fg** [comment, marked as answer]
Vale, perfecto.

Vamos a cambiar el if que tenemos:

```
if ($response->status() != 200 || false === $response->isDelivered() || false === $response->isCorrect()) {
    ToolBox::i18nLog()->warning('ticketbai-response-not-ok', ['%status%' => $response->mainErrorMessage()]);
    return false;
}
```

por algo así:

```
if ($response->status() != 200) {
    ToolBox::i18nLog()->warning('ticketbai-response-not-ok', ['%status%' => $response->mainErrorMessage()]);
    return false;
}

if (false === $response->isDelivered() || false === $response->isCorrect()) {
    ToolBox::i18nLog()->warning('ticketbai-response-not-ok', ['%status%' => $response->mainErrorMessage()]);
}

$qr = new Qr($this->ticketbai, $debugMode);
$this->tbaicodbar = $qr->ticketbaiIdentifier();
$this->tbaiurl = $qr->qrUrl();
$this->tbaisignature = $this->ticketbai->signatureValue();

if (empty($this->tbaicodbar) || empty($this->tbaiurl) || empty($this->tbaisignature)) {
ToolBox::i18nLog()->warning(('no hay datos de firma');
    return false;
}
```

Esta sería una propuesta ¿como la ve?

**Reply by arraintxo** [reply]
Parece que tiene sentido. Lo habéis probado ya?
